crates/extensions/c8y_firmware_manager/src/actor.rs

```diff
@@ -0,0 +1,723 @@
+            attempt: 1,
+        };
+
+        operation_entry.create_status_file(&self.config.firmware_dir)?;
+
+        self.publish_firmware_update_request(operation_entry, message_box)
+            .await?;
+
+        let operation_key = OperationKey::new(child_id, operation_id);
+        self.active_child_ops
+            .insert(operation_key.clone(), ActiveOperationState::Pending);
+
+        // Start timer
+        message_box
+            .send(SetTimeout::new(self.config.timeout_sec, operation_key).into())
+            .await?;
+
+        Ok(())
+    }
+
+    async fn handle_child_device_firmware_operation_response(
+        &mut self,
+        message: MqttMessage,
+        message_box: &mut FirmwareManagerMessageBox,
+    ) -> Result<(), FirmwareManagementError> {
+        let child_id = get_child_id_from_child_topic(&message.topic.name)?;
+
+        match FirmwareOperationResponse::try_from(&message) {
+            Ok(response) => {
+                if let Err(err) = self
+                    .handle_child_device_firmware_update_response(&response, message_box)
+                    .await
+                {
+                    self.fail_operation_in_cloud(
+                        &child_id,
+                        Some(response.get_payload().operation_id.as_str()),
+                        &err.to_string(),
+                        message_box,
+                    )
+                    .await?;
+                }
+            }
+            Err(err) => {
+                // Ignore bad responses. Eventually, timeout will fail an operation.
+                error!("Received a firmware update response with invalid payload for child {child_id}. Error: {err}");
+            }
+        }
+        Ok(())
+    }
+
+    async fn handle_child_device_firmware_update_response(
+        &mut self,
+        response: &FirmwareOperationResponse,
+        message_box: &mut FirmwareManagerMessageBox,
+    ) -> Result<(), FirmwareManagementError> {
+        let child_device_payload = response.get_payload();
+        let child_id = response.get_child_id();
+        let operation_id = child_device_payload.operation_id.as_str();
+        let received_status = child_device_payload.status;
+        info!("Firmware update response received. Details: id={operation_id}, child={child_id}, status={received_status:?}");
+
+        let operation_key = OperationKey::new(&child_id, &operation_id);
+        let current_operation_state = self.active_child_ops.get(&operation_key);
+
+        match current_operation_state {
+            Some(&ActiveOperationState::Executing) => {}
+            Some(&ActiveOperationState::Pending) => {
+                self.publish_c8y_executing_message(&child_id, message_box)
+                    .await?;
+                // TODO: Maybe cutting off as a function.
+                self.active_child_ops
+                    .insert(operation_key.clone(), ActiveOperationState::Executing);
```

Comment:
I'd do that in the `OperationStatus::Executing` arm of the `match received_status` statement below.

Reply:
In the end, both work the same from user perspective. But I think the current implementation is more natural. After sending executing message to c8y, the state becomes `OperationStatus::Executing`.